### ISSUE TITLE
Use structured polynomials to optimize sumcheck memory

### ISSUE THREAD

**Issue by lucasxia01** [AztecProtocol/barretenberg#1120]
We currently just allocate dyadic_circuit_size/2 size for all of the polynomials (including shifted ones!) during sumcheck, which is a huge amount of memory. In ClientIVC, we don't really care about this memory too much because its basically overshadowed by folding, which has 2 instances. However, for UltraHonk, sumcheck still is one of the peaks we care about. With lazy commitment key allocation, it no longer becomes the highest peak, but its still something to keep in mind in case it's a problem in WASM and it's possible that avoiding full polynomial allocations saves a little bit of time.

**Comment by fcarreiro** [comment]
Indeed handling shifts would also be very nice!

**Comment by lucasxia01** [comment]
Handling shifts shouldn't be any harder than normal polynomials? Like PartiallyEvaluatedMultivariates should include all entities which includes shifts.

**Comment by fcarreiro** [comment]
Hmm... ok depends on what "handling" means :) Right now we are allocating once for the original poly and again the same memory (approx) for the shifted one. I was hoping that by some clever means, you could just allocate once a PEM, for the original polynomial, and not for the shift. _That_ would be more difficult than just handling normal polys. In particular because you'd need to know at PEM construction time which polys are unshifted and which ones are shifted. Might not be a big deal depending on your Flavor, but certainly a bit more work than just doing what is done now :)

**Comment by lucasxia01** [comment]
No, that seems impossible because when we fold the PEMs, the shifted one will be completely different than the unshifted one.
